Index: libavformat/oggdec.c
===================================================================
--- libavformat/oggdec.c	(revision 21579)
+++ libavformat/oggdec.c	(working copy)
@@ -367,10 +367,10 @@

     ogg->curidx = idx;

-    if (os->header < 0){
+    if (!ogg->headers){
         int hdr = os->codec->header (s, idx);
+        os->header = os->seq;
         if (!hdr){
-            os->header = os->seq;
             os->segp = segp;
             os->psize = psize;
             ogg->headers = 1;
